tfhe/src/core_crypto/gpu/slice.rs

```diff
@@ -0,0 +1,192 @@
+use crate::core_crypto::gpu::vec::range_bounds_to_start_end;
+use crate::core_crypto::gpu::{CudaPtr, CudaStream};
+use crate::core_crypto::prelude::Numeric;
+use std::ffi::c_void;
+use std::marker::PhantomData;
+use tfhe_cuda_backend::cuda_bind::{cuda_memcpy_async_gpu_to_gpu, cuda_memcpy_async_to_cpu};
+
+#[derive(Debug, Copy, Clone)]
+pub struct CudaSlice<'a, T: Numeric> {
+    ptr: CudaPtr,
+    _len: usize,
+    _phantom_1: PhantomData<T>,
+    _phantom_2: PhantomData<&'a ()>,
+}
+
+#[derive(Debug)]
+pub struct CudaSliceMut<'a, T: Numeric> {
+    ptr: CudaPtr,
+    len: usize,
+    _phantom_1: PhantomData<T>,
+    _phantom_2: PhantomData<&'a mut ()>,
+}
+
+impl<'a, T> CudaSlice<'a, T>
+where
+    T: Numeric,
+{
+    /// # Safety
+    pub(crate) unsafe fn new(ptr: CudaPtr, len: usize) -> CudaSlice<'a, T> {
+        CudaSlice {
+            ptr,
+            _len: len,
+            _phantom_1: PhantomData,
+            _phantom_2: PhantomData,
+        }
+    }
```

Comment:
One last comment, we should probably add a Safety comment that says somethinglike `the ptr must be valid for reads for len * std::mem::size_of::<T> bytes on the cuda side` 

and for cuda slice mut `the ptr must be valid for read and write for len * std::mem::size_of::<T> bytes on the cuda side`

Reply:
Thanks a lot!